pkg/trace/writer/newwriter/sender.go

```diff
@@ -0,0 +1,458 @@
+	// in the sender.
+	recorder eventRecorder
+}
+
+// newSender creates a new sender, which uses the given HTTP client, URL and API key
+// to make requests using incoming payloads. The limit of outgoing concurrent connections
+// is defined by climit.
+func newSender(cfg *senderConfig) *sender {
+	return &sender{
+		cfg:    cfg,
+		climit: make(chan struct{}, cfg.maxConns),
+		list:   list.New(),
+	}
+}
+
+// Push pushes p onto the sender, to be written to the destination.
+func (q *sender) Push(p *payload) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	// enqueue the payload
+	q.enqueueLocked(p)
+
+	if !q.scheduled {
+		// no flush is scheduled; start one
+		q.scheduleFlushLocked(0)
+	}
+}
+
+// scheduleFlushLocked schedules the next flush using the given delay.
+func (q *sender) scheduleFlushLocked(delay time.Duration) {
+	switch {
+	case delay == 0:
+		go q.flush()
+	case q.timer == nil:
+		q.timer = time.AfterFunc(delay, q.flush)
+	default:
+		q.timer.Stop()
+		q.timer.Reset(delay)
+	}
+	q.scheduled = true
+}
+
+// flush drains and sends the entire queue. If anything comes in while flushing
+// or if some of the payloads fail to send as retriable, further follow up flushes
+// are scheduled.
+func (q *sender) flush() {
+	startTime := time.Now()
+
+	// we drain the queue, which is a blocking operation, meaning that
+	// new payloads are stopped from coming in while this happens; but
+	// it is a fast.
+	payloads := q.drainQueue()
+
+	// we send the payloads we've retrieved; while we do this, more payloads
+	// can join the list.
+	done, retries := q.sendPayloads(payloads)
+	if done > 0 {
+		q.recordEvent(eventTypeFlushed, &eventData{
+			count:    int(done),
+			duration: time.Since(startTime),
+		})
+	}
+
+	// we reassess the state of the list and check if further flushes need
+	// to be scheduled
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	switch {
+	case retries > 0:
+		// some sends failed as retriable; we need to back off a bit
+		q.attempt++
+		delay := backoffDuration(q.attempt)
+		q.scheduleFlushLocked(delay)
+	case q.list.Len() > 0:
+		// new items came in while we were flushing; schedule the next flush immediately.
+		q.scheduleFlushLocked(0)
+	default:
+		// everything is flushed
+		q.attempt = 0
+		q.scheduled = false
+	}
+}
+
+// drainQueue drains the entire queue and returns all the payloads that were in it.
+func (q *sender) drainQueue() []*payload {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	var payloads []*payload
+	for q.list.Len() > 0 {
+		v := q.list.Remove(q.list.Front())
+		payloads = append(payloads, v.(*payload))
+	}
+	q.size = 0
+	return payloads
+}
+
+// sendPayloads concurrently sends the given list of payloads. It returns
+// the number of payloads that were added back onto the queue to be retried
+// again later due to an error.
+func (q *sender) sendPayloads(payloads []*payload) (done, retries uint64) {
+	var wg sync.WaitGroup
+	for _, p := range payloads {
```

Comment:
> if a payload is ready to be sent, there's no reason not to send it out immediately (unless we're in backoff mode). I see the default flush interval is 0, so in theory we're sending them all out instantly, but still the concept of queueing it up for some flush seems unnecessary.

We _are_ sending them out immediately. There is no waiting. But I see what you mean. This is how the previous implementation worked too. It's absolutely natural to wonder: "why should we queue one payload onto a list, only to drain that list immediately and flush that payload after?". The answer is: code simplicity, readability and maintainability.

Even though it can seem like we are complicating a simple process, the code is now easier to test and reason about. Things happen in this order, and always follow the same code path:

1. Payload comes in
2. Payload gets queued (regardless of what's going on: queue is flushing, queue is not flushing, queue is active or backing off)
    * If a flush isn't scheduled already, one gets scheduled.
4. Queue get flushed at some point, then repeat.

Versus:

1. Payload comes in.
2. We check the state of the queue
    * Are we flushing? Enqueue on one code path.
    * Are we in backoff? Same as above.
    * Are we not flushing? Send on a different code path.

In short, I want to have only one method which flushes, not two: one that flushes immediately, and one that flushes by draining a queue. It just complicates things too much.

Looking at the [previous implementation](https://github.com/DataDog/datadog-agent/blob/a08cd3555c92780bcd6b580799e1e179aaec3c2d/pkg/trace/writer/sender.go), which works a bit like you're hinting at and is a great example of what you're saying, I strongly prefer this one, which even though it can make it seem like we are doing all this for nothing, only to immediately flush, the logic and code path is obvious, simple, easy to reason about, easy to modify/refactor/change and easy to test. 

Looking at resource usage, the cost is probably inexistent, or at the worst completely negligible.